src/h3lib/lib/algos.c

```diff
@@ -677,36 +673,162 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
+    for (int i = 0; i < numHexagons; i++) found[i] = 0;
+
+    // 4. Begin main loop. While the search hash is not empty do the following
+    while (numSearchHexes > 0) {
+        // Iterate through all hexagons in the current search hash, then loop
+        // through all neighbors and test Point-in-Poly, if point-in-poly
+        // succeeds, add to out and found hashes if not already there.
+        H3Index ring[7] = {0};
+        int currentSearchNum = 0;
+        int i = 0;
+        while (currentSearchNum < numSearchHexes) {
+            H3Index searchHex = search[i];
+            if (searchHex == 0) {  // Gap, skip it
+                i++;
+                continue;
+            }
+            H3_EXPORT(kRing)(searchHex, 1, ring);
+            for (int j = 1; j < 7; j++) {
+                if (ring[j] == 0) {
+                    continue;  // Skip if this was a pentagon and only had 5
+                               // neighbors
+                }
+
+                H3Index hex = ring[j];
+
+                // A simple hash to store the hexagon, or move to another place
+                // if needed. This MUST be done before the point-in-poly check
+                // since that's far more expensive
+                int loc = (int)(hex % numHexagons);
+                int loopCount = 0;
+                while (out[loc] != 0) {
+                    if (loopCount > numHexagons) {
+                        free(search);
+                        free(found);
+                        free(bboxes);
+                        return -1;
+                    }
+                    if (out[loc] == hex) break;  // Skip duplicates found
+                    loc = (loc + 1) % numHexagons;
+                    loopCount++;
+                }
+                if (out[loc] == hex) {
+                    continue;  // Skip this hex, already exists in the out hash
+                }
+
+                // Check if the hexagon is in the polygon or not
+                GeoCoord hexCenter;
+                H3_EXPORT(h3ToGeo)(hex, &hexCenter);
+                hexCenter.lat = constrainLat(hexCenter.lat);
+                hexCenter.lon = constrainLng(hexCenter.lon);
+
+                // If not, skip
+                if (!pointInsidePolygon(geoPolygon, bboxes, &hexCenter)) {
+                    continue;
+                }
+
+                // Otherwise set it in the output array
+                out[loc] = hex;
+
+                // Set the hexagon in the found hash
+                found[numFoundHexes] = hex;
+                numFoundHexes++;
+                // Wipe the current hex ring value back out
+                ring[j] = 0;
+            }
+            ring[0] = 0;
+            currentSearchNum++;
+            i++;
         }
+
+        // Swap the search and found pointers, copy the found hex count to the
+        // search hex count, and zero everything related to the found memory.
+        H3Index* temp = search;
```

Comment:
So I know ahead of time that you're going to be hard to convince, but I find using two separate arrays and then swapping them needlessly complicated here. Why not have a single `search` array, and treat it like a stack? 

- `numSearchHexes` tracks the last index
- A `pop` operation gets the index at `numSearchHexes`, zeroes its position in the array, and decrements `numSearchHexes`
- If the index hasn't been considered and passes point-in-poly, you push its neighbors into the array and increment `numSearchHexes` accordingly

Then you never have to iterate through `search`, you're always just popping the last index. You don't need a separate `found` array, because your new candidates get pushed onto `search`. This seems like it would be substantially simpler than the two-array version, and it only requires allocating 2x memory, not 3x like you have here (including the 1x allocated by the caller).

Reply:
The issue with that is that you're back to the depth-first search, and what will happen is something like this:

1. Get the outline of all of the geofences and put them into the search stack.
2. Take the last hexagon, check if it's in the geofence, then add all of its neighbors to the stack.
3. Use one of the neighbors and check it, then get all of its neighbors onto the stack and search it.
4. Your stack will likely grow larger than the entire array of memory you've allocated and you crash and burn.
5. Even if you solve that by going to a LinkedList approach like you originally proposed, the memory usage is going to be greater than what this is doing for any convex-like shape.
6. And when you finally finish filling the polygon, you're then going to do a lot of operations where you grab the next piece off of the stack and either confirm that it's already been added to the `out` array *or* you do a point-in-poly operation and confirm that it's out of bounds *or* you still have another array to hold H3Indexes that were point-in-poly rejected and check that and eliminate as you drop back the stack.
7. And many of these hexagons that got added to the stack are duplicated many times over since the check-and-remove step eliminates 1 hexagon while the get-neighbors-for-the-stack adds 6, so again, likely outcome is close to 6x memory usage here.

So, yes, I am hard to convince on this. :) But that's because I really do think the other approach will not actually save any memory or CPU time.

Reply:
OK, that makes sense. It's primarily the complexity I'm responding to here. I'm not sure I'm following your number 6 here, since as far as I can tell you're still doing the same number of operations on the same number of hexagons as you would with a stack, but you're probably right that DFS isn't bounded to the max polyfill size, particularly if you aren't deduping when you add to the stack.